lib/bridge/IrcHandler.js

```diff
@@ -538,15 +538,90 @@ IrcHandler.prototype.onMode = Promise.coroutine(function*(req, server, channel,
                                                 mode, enabled, arg) {
     let privateModes = ["k", "i", "s"];
 
-    if (privateModes.indexOf(mode) === -1) {
-        return; // ignore everything but privateModes
-    }
     req.log.info(
         "onMode(%s) in %s by %s (arg=%s)",
         (enabled ? ("+" + mode) : ("-" + mode)),
         channel, by, arg
     );
 
+    if (privateModes.indexOf(mode) !== -1) {
+        yield this._onPrivateMode(req, server, channel, by, mode, enabled, arg);
+        return;
+    }
+
+    // Bridge operator modes to power levels
+    let modeToPower = server.getModePowerMap();
+
+    if (Object.keys(modeToPower).indexOf(mode) === -1) {
+        // Not an operator power mode
+        return;
+    }
+
+    const nick = arg;
+
+    const matrixRooms = yield this.ircBridge.getStore().getMatrixRoomsForChannel(server, channel);
+
+    if (matrixRooms.length === 0) {
+        req.log.info("No mapped matrix rooms for IRC channel %s", channel);
+        return;
+    }
+
+    const bridgedClient = this.ircBridge.getClientPool().getBridgedClientByNick(
+        server, nick
+    );
+
+    if (!bridgedClient) {
+        req.log.info("No bridged client for %s", nick);
+        return;
+    }
+
+    if (!bridgedClient.unsafeClient) {
+        req.log.info(`Bridged client for ${nick} has no IRC client.`);
+        return;
+    }
+    const chanData = bridgedClient.unsafeClient.chanData(channel);
+
+    if (!(chanData && chanData.users)) {
+        req.log.error(`No channel data for ${channel}`);
+        return;
+    }
+
+    const userPrefixes = chanData.users[nick];
+    const userPowers = [];
+    userPrefixes.split('').forEach(
+        prefix => {
+            const m = bridgedClient.unsafeClient.modeForPrefix[prefix];
+
+            if (modeToPower[m] !== undefined) {
+                userPowers.push(modeToPower[m]);
+            }
+        }
+    );
+
+    // Sort the userPowers for this user in descending order
+    // and grab the highest value at the start of the array.
+    let level = 0;
+    if (userPowers.length > 0) {
```

Comment:
If `userPowers` is empty, we will give this user a power level of `0`. What we probably mean instead is to default to `users_default` in the `m.room.power_levels` event which is not the same thing.